[reland] Fix estimate_nccl_collective_runtime

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119229
* #119104
* #119103
* __->__ #118986
* #119102

`estimate_nccl_collective_runtime` has been broken and the errors have been silently swallowed by inductor. This PR:
- Fixes the issues described in https://github.com/pytorch/pytorch/issues/118497.
- Adds white-box testing so future issues can be surfaced in tests.
- Add support for native funcol IRs.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Commit:
Fix estimate_nccl_collective_runtime

[ghstack-poisoned]

Commit:
Update on "Fix estimate_nccl_collective_runtime"


`estimate_nccl_collective_runtime` has been broken and the errors have been silently swallowed by inductor. This PR:
- Fixes the issues described in https://github.com/pytorch/pytorch/issues/118497.
- Adds white-box testing so future issues can be surfaced in tests.
- Add support for native funcol IRs.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

wconstab commented on test/inductor/test_snode_runtime.py:
```diff
@@ -165,6 +172,132 @@ def f(x):
+            metrics.reset()
+            torch._dynamo.optimize(count_bytes_inductor)(fn)(*inps)
+            found_collective = False
+            for snode, runtime in metrics.node_runtimes:
```
cool, is this an estimate from the scheduler for every node, (iirc there were some lo-fi estimates like eltwise was cheap and gemm is expensive, but I didn't remember if it was calibrated (is the collective estimator predicting actual time intervals?)

wconstab replied:
nvm see it below. time in ns.

wconstab commented on test/inductor/test_snode_runtime.py:
```diff
@@ -165,6 +172,132 @@ def f(x):
+
+    def test_all_gather_into_tensor(self):
+        def fn(x):
+            rs = _c10d.all_gather_into_tensor(
```
hm, what's this doing- these are not functional collectives, so I wouldn't have expected them to work with wait_tensor.  or, is that counting on dynamo doing a rewrite?

yifuwang replied:
Hey @wconstab, these are dispatcher ops. For both the existing and new funcol, "tensor wrapping" happens inside dispatcher calls.

The reason I resorted to dispatcher ops for this test is that there's currently some issues with tracing some of the new funccols. This is okay for this test since it focuses on the IR. I will add tracing tests as I resolve the aforementioned issues.

wconstab commented on torch/_inductor/scheduler.py:
```diff
@@ -581,6 +583,16 @@ def get_estimated_runtime(self) -> float:
             # default to no reordering based on runtime
             return 0
 
+        # Collective kernels
```
you had to move this higher up since it was getting bypassed previously?

yifuwang replied:
Yes. Previously this part always failed and returned 0 because for collective kernels, the dtype obtained above is always `None`.

```python
        try:
            gpu_memory_bandwidth = get_gpu_dram_gbps()
            gpu_flops = get_device_tflops(dtype) * 10**12
        except Exception:
            return 0
```

wconstab commented on torch/_inductor/utils.py:
```diff
@@ -1258,3 +1258,15 @@ def pass_execution_and_save(func, gm, msg):
+def is_collective(node):
+    from . import ir
+
+    return isinstance(node, ir.CollectiveKernel) or type(node) == ir._CollectiveKernel
```
this is subtle, something related to new types of collectives being lowered differently?

yifuwang replied:
You mean `type(node) == ir._CollectiveKernel`? In the new IRs, `ir._WaitKernel` derives from `ir._CollectiveKernel` to share some stuff needed for the cpp codegen.

yf225 commented on torch/csrc/distributed/c10d/FakeProcessGroup.hpp:
```diff
@@ -9,6 +9,12 @@ class FakeWork : public Work {
+    auto fut = c10::make_intrusive<c10::ivalue::Future>(c10::NoneType::get());
+    fut->markCompleted();
+    return fut;
+  }
```
dumb q: is this part related to the PR?

yf225 reviewed (approved):
Thanks a lot for fixing this @yifuwang !

yifuwang commented:
@pytorchbot merge

Commit:
Update on "Fix estimate_nccl_collective_runtime"


`estimate_nccl_collective_runtime` has been broken and the errors have been silently swallowed by inductor. This PR:
- Fixes the issues described in https://github.com/pytorch/pytorch/issues/118497.
- Adds white-box testing so future issues can be surfaced in tests.
- Add support for native funcol IRs.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Commit:
Update on "Fix estimate_nccl_collective_runtime"


`estimate_nccl_collective_runtime` has been broken and the errors have been silently swallowed by inductor. This PR:
- Fixes the issues described in https://github.com/pytorch/pytorch/issues/118497.
- Adds white-box testing so future issues can be surfaced in tests.
- Add support for native funcol IRs.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Commit:
Update on "Fix estimate_nccl_collective_runtime"


`estimate_nccl_collective_runtime` has been broken and the errors have been silently swallowed by inductor. This PR:
- Fixes the issues described in https://github.com/pytorch/pytorch/issues/118497.
- Adds white-box testing so future issues can be surfaced in tests.
- Add support for native funcol IRs.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge

atalman commented:
@pytorchmergebot revert -c ghfirst -m "Breaks internal tests"

Commit:
Update on "Fix estimate_nccl_collective_runtime"


`estimate_nccl_collective_runtime` has been broken and the errors have been silently swallowed by inductor. This PR:
- Fixes the issues described in https://github.com/pytorch/pytorch/issues/118497.
- Adds white-box testing so future issues can be surfaced in tests.
- Add support for native funcol IRs.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Commit:
Update on "Fix estimate_nccl_collective_runtime"


`estimate_nccl_collective_runtime` has been broken and the errors have been silently swallowed by inductor. This PR:
- Fixes the issues described in https://github.com/pytorch/pytorch/issues/118497.
- Adds white-box testing so future issues can be surfaced in tests.
- Add support for native funcol IRs.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Commit:
Update on "Fix estimate_nccl_collective_runtime"


`estimate_nccl_collective_runtime` has been broken and the errors have been silently swallowed by inductor. This PR:
- Fixes the issues described in https://github.com/pytorch/pytorch/issues/118497.
- Adds white-box testing so future issues can be surfaced in tests.
- Add support for native funcol IRs.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kwen2501 commented:
Hi @yifuwang sorry just get to notice that we have this utility. (Which would be indeed helpful for comm scheduling)
I wonder if there would be enough time for pytorch to read in a matrix dumped by NCCL during init?
I previously opened a PR towards NCCL to dump its tuning matrix containing the bandwidth and latency numbers. See https://github.com/NVIDIA/nccl/issues/830.
Since it is prepared **AFTER** NCCL performs topology detection, its bandwidth/latency numbers are most accurate.
In PyTorch, I'd think we would avoid any topology detection (too low level, also lack of APIs). 
Also, the tuning matrix of NCCL would match how NCCL selects algorithm and protocol on the specific platform the job is run on.

yifuwang commented:
@pytorchbot merge
